src/services/item/entities/Item.ts

```diff
@@ -100,3 +107,46 @@ export class Item extends BaseEntity implements GraaspItem {
   // @ManyToMany(() => ItemCategory, (iC) => iC.item)
   // categories: ItemCategory[];
 }
+
+// all sub-item types defined using a specific variant of the `ItemType` enumeration
+export type AppItem = Item<typeof ItemType.APP>;
+export type DocumentItem = Item<typeof ItemType.DOCUMENT>;
+export type EtherpadItem = Item<typeof ItemType.ETHERPAD>;
+export type FolderItem = Item<typeof ItemType.FOLDER>;
+export type H5PItem = Item<typeof ItemType.H5P>;
+export type EmbeddedLinkItem = Item<typeof ItemType.LINK>;
+export type LocalFileItem = Item<typeof ItemType.LOCAL_FILE>;
+export type S3FileItem = Item<typeof ItemType.S3_FILE>;
+export type ShortcutItem = Item<typeof ItemType.SHORTCUT>;
+
+// Typeguard definitons that help to narrow the type of an item to one of the specific item types
+// typeguard are used because a class does not support discriminated union
+// (which are the recommented way of achieving what we have with the item type definiton)
+// One benefit of using the typeguards is that it shortens the code when checking for an item type:
+// Before:
+//          if (item.type === ItemType.FOO) {
+//            // item.extra was not correctly typed to the correcponding extra type
+//            // do something, but you will have to cast the item.extra ... not ideal
+//          }
+//
+// Now:
+//          if (isFooItem(item)) {
+//            // now item.extra is of the mapped type, i.e: FooItemExtra
+//            // do some smarter things without needing to cast the extra
+//          }
+export const isAppItem = (item: Item<ItemTypeKeys>): item is AppItem => item.type === ItemType.APP;
+export const isDocumentItem = (item: Item<ItemTypeKeys>): item is DocumentItem =>
+  item.type === ItemType.DOCUMENT;
+export const isEtherpadItem = (item: Item<ItemTypeKeys>): item is EtherpadItem =>
+  item.type === ItemType.ETHERPAD;
+export const isFolderItem = (item: Item<ItemTypeKeys>): item is FolderItem =>
+  item.type === ItemType.FOLDER;
+export const isH5PItem = (item: Item<ItemTypeKeys>): item is H5PItem => item.type === ItemType.H5P;
+export const isEmbeddedLinkItem = (item: Item<ItemTypeKeys>): item is EmbeddedLinkItem =>
+  item.type === ItemType.LINK;
+export const isLocalFileItem = (item: Item<ItemTypeKeys>): item is LocalFileItem =>
+  item.type === ItemType.LOCAL_FILE;
+export const isS3FileItem = (item: Item<ItemTypeKeys>): item is S3FileItem =>
+  item.type === ItemType.S3_FILE;
+export const isShortcutItem = (item: Item<ItemTypeKeys>): item is ShortcutItem =>
+  item.type === ItemType.SHORTCUT;
```

Comment:
Are these type guards necessary? I thought the change would allow the typescript compiler to automatically pickup the extra given the disciminated union on `item.type`? IMO type guards are in essence equivalent to casts, since we basically tell the compiler "trust me bro, if this check passes then we promise it's the `is` type".

I.e. does this mean that writing the following in the code somewhere does not provide the appropriate extra typings (as it does in the front-end)?
```ts
if (item.type === ItemType.DOCUMENT) {
    typeof item.extra // <-- should be DocumentItemExtra but is ?
}
```

When typescript recognizes a type union you shouldn't need typeguards: see https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types

If we still need these typeguards, I think we did not gain additional safety, i.e. we could have written the typeguard before as well and from the type system guarantees they would be the same (either previously the casts, or now the type guards, provide the actual "guarantee"), unlike in the SDK where we currently have "true" union type discrimination.

If we don't, then you should drop them as they may introduce type errors in the future. Consumers should be able to expect that checking against `item.type` correctly narrows the extra type.

Reply:
You are correct, checking `item.type === "folder"` does not narrow the generic as outlined in [this SO post](https://stackoverflow.com/questions/68897217/types-narrowing-with-generics). This really sucks in my opinion as I would have like for it to work in the described way. 

The issue is that we can not define a union type at the level of the class ...